proposals/p1885.md

```diff
@@ -0,0 +1,866 @@
+    }
+  }
+}
+
+fn IterateBidirectional.Reversed[self: Self]() -> ReverseProxy(Self) {
+  return {.container = self};
+}
+
+/// Usage
+class MyContainer {
+  fn Create(...) { ... }
+  impl as Iterate ... { ... }
+  impl as IterateBidirectional ... { ... }
+}
+var container: MyContainer = MyContainer.Create(...);
+for (v: container.(Iterate.ElementType) in container.Reversed()) {...}
+```
+
+#### Interoperability with C++ types and containers
+
+The cursor approach deviates from the C++ iterator approach, requiring some
+adaptation.
+
+Given the current state of the design, C++ interoperability is yet to be
+defined, and suggestions have been highlighted in the
+[Future work](#future-work) section.
+
+## Future work
+
+### `for` with mutable values
+
+Supporting mutable values could be achieved using a second interface that
+provides a proxy or view for the data, exposing an `Iterate` interface with
+`ElementType*`
+
+```carbon
+// Object implementing `MutableIterate` returned by a function
+for (p: T* in my_range.AddrRange()) {}
+```
+
+The example `MutableIterate` interface below acts as a proxy for the collection,
+and implements our `Iterate` interface.
+
+```carbon
+interface MutableIterate {
+  impl as Iterate;
+  alias ElementType = Iterate.ElementType;
+  let ProxyType:! Iterate where .ElementType = ElementType*
+      and .CursorType is ImplicitAs((Self as Iterate).CursorType);
+  fn AddrRange[addr self: Self*]() -> ProxyType;
+}
+
+class MyIntContainer {
+  //...
+  impl as Iterate ...
+  external impl as MutateIterate where .ProxyType = Self {
+    ...
+  }
+}
+```
+
+### Speculative mixin usage
+
+Mixins are currently in early design stages. This section highlights possible
+uses speculating on the final design. Some may include:
+
+-   Improved semantics for views compared to getter methods: no direct side
+    effect from using the view
+-   Facilitate code reuse, compared to reimplementing an interface
+-   Direct access to `self`, limiting needs for pointers and address resolution
+
+#### Mutable values
+
+Named mixins, used as programmable members, could expose a view with mutable
+values:
+
+```carbon
+// Using a named mixin to expose mutable values
+for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
+```
+
+Used in this way, this clarifies that this a view of the data, like an
+attribute, with no direct side effects, when compared to a method.
+
+#### Alternative views
+
+Similarly, mixins could allow supporting other views into the container.
+
+For example, maps could expose a view with "key" and "value" only, in addition
+to the default (key, value) tuple.
+
+```carbon
+// Using a named mixin to expose mutable values
+for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
+```
+
+And "reversed" view could also be used for reverse iteration:
+
+```carbon
+// Reversing
+for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
+```
+
+### Large elements and Optional
```

Comment:
@chandlerc I feel like we had a conversation about this and r-value representations. Could you recall where we ended up with this?